presto-cache/src/main/java/com/facebook/presto/cache/LocalRangeCacheManager.java

```diff
@@ -0,0 +1,461 @@
+import static java.nio.file.StandardOpenOption.CREATE_NEW;
+import static java.util.Objects.requireNonNull;
+import static java.util.UUID.randomUUID;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+// 3 major TODOs for this class:
+// TODO: Make cache eviction based on cache size rather than file count; add evict count stats to CacheStats as well.
+// TODO: Make cache state persistent on disk so we do not need to wipe out cache every time we reboot a server.
+// TODO: File merge and inflight requests memory control.
+@SuppressWarnings("UnstableApiUsage")
+public class LocalRangeCacheManager
+        implements CacheManager
+{
+    private static final Logger log = Logger.get(LocalRangeCacheManager.class);
+
+    private static final String EXTENSION = ".cache";
+
+    private final ExecutorService cacheFlushExecutor;
+    private final ExecutorService cacheRemovalExecutor;
+
+    // a mapping from remote file `F` to a range map `M`; the corresponding local cache file for each range in `M` represents the cached chunk of `F`
+    private final Map<Path, CacheRange> persisted = new ConcurrentHashMap<>();
+    // a localing cache only to control the lifecycle of persisted
+    private final Cache<Path, Boolean> cache;
+
+    // stats
+    private final CacheStats stats;
+
+    // config
+    private final Path directory;
+    private final long maxInflightBytes;
+
+    public LocalRangeCacheManager(CacheConfig cacheConfig, CacheStats stats, ExecutorService cacheFlushExecutor, ExecutorService cacheRemovalExecutor)
+    {
+        requireNonNull(cacheConfig, "directory is null");
+        this.cacheFlushExecutor = cacheFlushExecutor;
+        this.cacheRemovalExecutor = cacheRemovalExecutor;
+        this.cache = CacheBuilder.newBuilder()
+                .maximumSize(cacheConfig.getMaxCachedEntries())
+                .expireAfterAccess(cacheConfig.getCacheTtl().toMillis(), MILLISECONDS)
+                .removalListener(new CacheRemovalListener())
+                .recordStats()
+                .build();
+        this.stats = requireNonNull(stats, "stats is null");
+        this.directory = new Path(cacheConfig.getBaseDirectory());
+        checkArgument(cacheConfig.getMaxInMemoryCacheSize().toBytes() >= 0, "maxInflightBytes is negative");
+        this.maxInflightBytes = cacheConfig.getMaxInMemoryCacheSize().toBytes();
+
+        File target = new File(directory.toUri());
+        if (!target.exists()) {
+            try {
+                Files.createDirectories(target.toPath());
+            }
+            catch (IOException e) {
+                throw new PrestoException(GENERIC_INTERNAL_ERROR, "cannot create cache directory " + target, e);
+            }
+        }
+        else {
+            File[] files = target.listFiles();
+            if (files == null) {
+                return;
+            }
+
+            this.cacheRemovalExecutor.submit(() -> Arrays.stream(files).forEach(file -> {
```

Comment:
I'm not entirely sure how the file system internal works, but it seems more straightforward to just delete the directory recursively and recreate the directory.

Reply:
It takes about an hour to delete the directory recursively. So that's why we moved to a background job.

Reply:
Oh okay. Our internal HDFS mutation has a GC service and directory deletion returns immediately and that's why it leads me to this. Thanks for explanation.